Create pull request/issue#644

Jasielprogramador opened the pull request:
I think everything is good to go, implemented things:
1) Each card on the LoyaltyCardEditActivity has a archive button, black when unarchived and white when archived
2) After long press in 1 or more cards in the overflow menu you can archive/unarchive cards
3) Archive icon in archived cards

Jasielprogramador commented:
I'm open to hearing your opinion about how i modified the code cause i had to do a lot of changes and implement lot of things

Altonss commented:
I'm really excited to test this new feature, once an apk is available I will test it out!

TheLastProject commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -162,6 +163,22 @@ public boolean onActionItemClicked(ActionMode inputMode, MenuItem inputItem) {
             }
+            else if(inputItem.getItemId() == R.id.action_archive_unarchive){
+                for (LoyaltyCard loyaltyCard : mAdapter.getSelectedItems()) {
+                    Log.e(TAG, "Archiving card: " + loyaltyCard.id);
```
e in `Log.e` means error. It's not an error. `Log.d` would make more sense

TheLastProject replied:
I see I messed that up for card deletion too, but better to fix that too than to make it wrong twice :)

TheLastProject commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -519,16 +538,27 @@ public boolean onOptionsItemSelected(MenuItem inputItem) {
+            CheckBox showArchived = (CheckBox) customLayout.findViewById(R.id.checkBox_archived);
+
             ch.setChecked(mOrderDirection == DBHelper.LoyaltyCardOrderDirection.Descending);
+            showArchived.setChecked(mArchiveFilter == DBHelper.LoyaltyCardArchiveFilter.Archived);
```
If we do the All/Archived/Unarchived, this should probably be All

TheLastProject commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -519,16 +538,27 @@ public boolean onOptionsItemSelected(MenuItem inputItem) {
                 }
+
+                if(showArchived.isChecked())
+                    setArchivedSort(DBHelper.LoyaltyCardArchiveFilter.Archived);
```
If we do the All/Archived/Unarchived, this should probably be All

TheLastProject commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -585,6 +615,22 @@ private void setSort(DBHelper.LoyaltyCardOrder order, DBHelper.LoyaltyCardOrderD
+                getString(R.string.sharedpreference_sort),
+                Context.MODE_PRIVATE);
+        SharedPreferences.Editor sortPrefEditor = sortPref.edit();
+        sortPrefEditor.putString(getString(R.string.sharedpreference_sort_filter), filter.name());
```
I think this line should just be part of setSort, and this whole function is unnecessary

TheLastProject replied:
And I'd rather not name this "filter", that's too generic. It's also not really related to sorting, we're just putting it in the same menu as it seems to fit nicely. I'd say `sharedpreference_show_archived` would be better.

TheLastProject commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -162,6 +163,22 @@ public boolean onActionItemClicked(ActionMode inputMode, MenuItem inputItem) {
+                    }
+                    else{
+                        num = 0;
+                    }
```
I think this will be very confusing for people, they select multiple cards and instead of them all archiving or unarchiving they'll flip state around.

TheLastProject commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -685,6 +731,10 @@ private void toggleSelection(int inputPosition) {
             MenuItem editItem = mCurrentActionMode.getMenu().findItem(R.id.action_edit);
+
+            MenuItem archiveItem = mCurrentActionMode.getMenu().findItem(R.id.action_archive_unarchive);
+            archiveItem.setVisible(true);
```
I think it makes more sense to turn these into 2 menu items. One for archiving, one for unarchiving.

That way you can:
1. Show only archiving if all items aren't archived
2. Show only unarchiving if all items are archived
3. Show both if the items are in a mixed state

That also prevents confusion with state flipping around.

Jasielprogramador replied:
The thing is that i don't really get what do you want, i mean you have the checkbox in order to see the archived and all or only the unarchived ones

Jasielprogramador replied:
The thing is that i don't really get what do you want, i mean you have the checkbox in order to see the archived and all or only the unarchived ones

Jasielprogramador replied:
Oh yeah but i already implemented that the only thing if you select more than one you get the options Archive All and Unarchive all

TheLastProject commented on app/src/main/java/protect/card_locker/importexport/CatimaImporter.java:
```diff
@@ -321,6 +321,15 @@ private void importLoyaltyCard(Context context, SQLiteDatabase database, CSVReco
+        try {
+            archiveStatus = CSVHelpers.extractInt(DBHelper.LoyaltyCardDbIds.ARCHIVE_STATUS, record, false);
+        } catch (FormatException _e) {
+            // This field did not exist in versions 0.28 and before
```
2.16.3, not 0.28 :)

TheLastProject commented on app/src/main/res/menu/card_longclick_menu.xml:
```diff
@@ -30,4 +30,10 @@
+        android:icon="@drawable/ic_baseline_archive_24"
+        android:title="@string/archive_unarchive"
+        android:titleCondensed="@string/archive_unarchive"
+        app:showAsAction="ifRoom"/>
```
This will put archiving after deleting. I think it makes more sense to keep deleting as the last item.

TheLastProject reviewed (changes requested):
I haven't ran the code yet (and I want to wait a bit until the database migrations aren't in a broken state so I don't have to switch back and forth builds all the time) but it is looking like a good start. Do have some notes though :)

Jasielprogramador commented on app/src/main/java/protect/card_locker/DBHelper.java:
```diff
@@ -606,13 +636,20 @@ public static Cursor getLoyaltyCardCursor(SQLiteDatabase database, String filter
+            archiveString.append(" WHERE " + LoyaltyCardDbIds.TABLE + "." + LoyaltyCardDbIds.ARCHIVE_STATUS + " = 0 ");
+        else{
+            archiveString.append(" ");
+        }
```
I mean i could do that but i think that it complicates the issue, cause i would have to add another checkbox or something so that only the archived cards would be shown. I don't mind adding that but i don't really think that it would be practical, right now the user can see all the cards and distinguish which one is archived and which one is not just by seeing the icon of archived in the card, i think this way i just the easiest way cause the user would be able to "restore" (unarchive) the card that he wants. Tell me what do you prefer or if you have a better idea, i'm all ears.

TheLastProject replied:
I meant more to use clearer naming of the LoyaltyCardArchiveFilter enum. Don't need an extra checkbox or anything.

How about:
LoyaltyCardArchiveFilter.All for all cards
LoyaltyCardArchiveFilter.Active for all cards not archived

Jasielprogramador replied:
Ohh i see yeah that seems a lot easier to understand gonna change it right now

TheLastProject replied:
Looks better, but there's no need to use a StringBuilder if it's just a plain string.

I'd say just move it into the rawQuery like filter is:
```
(filter.trim().isEmpty() ? " " : " AND " + LoyaltyCardDbFTS.TABLE + " MATCH ? ") +
```

Jasielprogramador commented on app/src/main/java/protect/card_locker/LoyaltyCardCursorAdapter.java:
```diff
@@ -355,6 +358,14 @@ public void setIconBackgroundColor(int color) {
+            } else {
+                mArchived.setVisibility(View.GONE);
+            }
+        }
```
The way i think about this is that a archived card can also be a starred card, this way if you unarchive it it would show that is starred. If you put a enum like this:

```java 
public enum LoyaltyCardStateIcon{
Archived,
Starred,
None}
```
A card could not be archived and starred  at the same time

TheLastProject replied:
Makes sense, then turning `toggleStar` into `showCardIcons(boolean starred, boolean archived, boolean colorByTheme)` would probably be better. Because probably the archive icon will also need the same foreground and background handling as the star icon.

Jasielprogramador commented on app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:
```diff
@@ -140,6 +144,7 @@ public class LoyaltyCardEditActivity extends CatimaAppCompatActivity {
 
     int loyaltyCardId;
     boolean updateLoyaltyCard;
+    boolean duplicateFromLoyaltyCardId;
```
Sure sorry my bad i had a little bit of a mess with the branches

Jasielprogramador commented on app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:
```diff
@@ -1314,13 +1322,43 @@ private void doSave() {
         if (updateLoyaltyCard) {
             getMenuInflater().inflate(R.menu.card_update_menu, menu);
+            LoyaltyCard loyaltyCard = DBHelper.getLoyaltyCard(mDatabase, loyaltyCardId);
+            archived = loyaltyCard.archiveStatus != 0;
```
Ok got it yeah i think the same way, dw i got this hahhaha

Jasielprogramador commented:
All the asked code was modified, implemented things:

1) Now, is easier for the user to archive a card, if he enters inside the `LoyaltyCardViewActivity` class in the overflow menu of the card you will now have the archive/unarchive (if card is archived will show unarchive option and otherwise) and delete options that where previously imnplemented in ´LoyaltyCardEditActivity`

2) Now, if the user long presses 1 card  if that card is archived will show unarchive option in the overflow menu and otherwise, at the same time if he presses more than 1 card the user will now see 2 options in the overflow menu, Archive all and unarchive all.

3) In the sort menu the user will find the option to see the archived cards, if checked the archived cards will be shown as well with the unarchived ones, if not only the unarchived ones will be shown.

4) The archived cards that are shown with the previous check will be differenced cause the archived ones have the archive icon in them.

TheLastProject commented on app/src/main/java/protect/card_locker/LoyaltyCardCursorAdapter.java:
```diff
@@ -335,20 +339,41 @@ public void toggleStar(boolean enable, boolean colorByTheme) {
+            }
+
+            mArchivedBorder.invalidate();
+            mArchivedBackground.invalidate();
```
I'm not a huge fan of:

```
if (dark) {
  A
} else {
  B
}

C.invalidate()

if (dark) {
  D 
} else {
  E
}

F.invalidate()
```

I'd find it much simpler to use:
```
if (dark) {
  A
  D
} else {
  B
  E
}

C.invalidate()
F.invalidate()
```

Although this is kinda a nitpick

TheLastProject commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -519,16 +539,27 @@ public boolean onOptionsItemSelected(MenuItem inputItem) {
                 }
+                else if(!ch.isChecked() && !showArchived.isChecked()){
+                    setSort(loyaltyCardOrders.get(currentIndex.get()), DBHelper.LoyaltyCardOrderDirection.Ascending, DBHelper.LoyaltyCardArchiveFilter.Active);
+                }
```
I think this whole if/else can be simplified as something like:

```
setSort(
  loyaltyCardOrders.get(currentIndex.get()),
  ch.isChecked() ? DBHelper.LoyaltyCardOrderDirection.Descending : DBHelpder.LoyaltyCardOrderDirection.Ascending,
  showArchived.isChecked() ? DBHelper.LoyaltyCardArchiveFilter.All : DBHelper.LoyaltyCardArchiveFilter.Active
)
```

Given there's now 2 options, `ch` (short for checkbox) should be renamed because it is no longer the only checkbox. Something like `sortReversed` sounds good.

TheLastProject commented:
I'm not sure why you requested another review, there are still a few open comments and it's still hard to review without rebase :)

Jasielprogramador commented:
Oh sorry my bad

Jasielprogramador commented:
Ok well i rebased the branch and i changed everything that was asked or i believe so :D

Altonss commented:
> Ok well i rebased the branch and i changed everything that was asked or i believe so :D

Github still says that there are conflicts that need to be resolved :thinking:

Altonss reviewed (changes requested):
The tests are not passing, it seems like the DBHelper.insertLoyaltyCard() is failing in the test files, so maybe these need to be updated?

Jasielprogramador commented:
Ok i resolved the conflicts but the thing is that i get that DBHelper.insertLoyaltyCard(...) is not applicable, because i changed this method due to i added a new column into the card table in order to save if a card is archived or not, how can i change this so that it passes the tests?

Altonss commented:
> Ok i resolved the conflicts but the thing is that i get that DBHelper.insertLoyaltyCard(...) is not applicable, because i changed this method due to i added a new column into the card table in order to save if a card is archived or not, how can i change this so that it passes the tests?

You need to change the files in the test folder. I found these files using this method:
app/src/test/java/protect/card_locker/LoyaltyCardCursorAdapterTest.java
app/src/test/java/protect/card_locker/ImportExportTest.java
app/src/test/java/protect/card_locker/LoyaltyCardViewActivityTest.java
app/src/test/java/protect/card_locker/DatabaseTest.java
app/src/test/java/protect/card_locker/MainActivityTest.java
app/src/test/java/protect/card_locker/ImportURITest.java
app/src/test/java/protect/card_locker/ImportExportTest.java

Jasielprogramador commented:
Hello i have an issue, the thing is that the archive column does not exist in the database used in DatabaseTest using this:

```[java]
 mActivity = Robolectric.setupActivity(MainActivity.class);
 mDatabase = TestHelpers.getEmptyDb(mActivity).getWritableDatabase();
```

How  can i change the fields of this database???

TheLastProject commented:
The failing test is "databaseUpgradeFromVersion1", so your code works fine for new installation, but is failing for upgrades.

You need to:
1. Upgrade DATABASE_VERSION from 14 to 15 (because you're adding a new "database version")
2. In onUpgrade, add your ARCHIVE_STATUS column when going from version 14 to 15:
```
        if (oldVersion < 15 && newVersion >= 15) {
            db.execSQL("ALTER TABLE " + LoyaltyCardDbIds.TABLE
                    + " ADD COLUMN " + LoyaltyCardDbIds.ARCHIVE_STATUS + " INTEGER DEFAULT '0' ");
        }
```

(I believe in more modern Android with "Room" this shouldn't be necessary, but Catima is based on some older Android database management where we're still responsible for this ourselves)

Jasielprogramador commented:
Hello there, now i pass the tests but it fails in the spotbugs, what is that? Btw thanks so much for the help you two and sorry for the bother

TheLastProject commented:
No worries, I'm glad you're still working on this and haven't given up after all the changes I requested :)

SpotBugs is a little tool that finds code that could contain bugs. It doesn't necessarily have to be buggy code, but it helps reduce the changes of subtle bugs.

If you go to the "Summary" page for the build you'll see "test-results" in Artifacts. It contains the spotbugs results in `spotbugs/release.html`.

Altonss commented:
> No worries, I'm glad you're still working on this and haven't given up after all the changes I requested :)
> 
> SpotBugs is a little tool that finds code that could contain bugs. It doesn't necessarily have to be buggy code, but it helps reduce the changes of subtle bugs.
> 
> If you go to the "Summary" page for the build you'll see "test-results" in Artifacts. It contains the spotbugs results in `spotbugs/release.html`.

Here you have the issue:
![image](https://user-images.githubusercontent.com/66519591/172836545-de305b34-d99d-479b-a3fa-d36c7cc23a81.png)

TheLastProject commented:
The easiest fix for that is to probably just switch to if statements, that seems to be what most of the other code in the project uses too:

https://github.com/CatimaLoyalty/Android/blob/6d134009beb03d2143fbee76402a325d62414364/app/src/main/java/protect/card_locker/ManageGroupActivity.java#L168-L179

Jasielprogramador commented:
Ok finally let's gooooooooooo, i made the spotbugs happy hahhahah

TheLastProject commented:
It's great to see all the tests are green. I'll do some more testing and reviewing probably tomorrow or this weekend ;)

TheLastProject commented:
Okay, some quick UI testing:

1. When you archive a card, it disappears completely. This is going to confuse some people. I'd much prefer if "show archived" in "Sort by" defaults to true
2. When you press "Hide details", the archive icon disappears (it should stay visible like the star icon)
3. Archived cards are shown in the same space as regular cards, they should probably always be sorted to the bottom because they're "less important". So the order should be as followed: Starred -> Regular -> Starred Archived -> Regular Archived
4. If you select multiple cards you always get "Archive all" and "Unarchive all". In toggleSelection you get a `mAdapter.getSelectedItems()` variable, you probably want to have logic like this (pseudocode, untested) to only show the relevant option:
```
hasArchived = false
hasUnarchived = false

for (card in mAdapter.getSelectedItems()) {
  if (card.archiveStatus == 1) {
    hasArchived = true;
  } else {
    hasUnarchived = true;
  }

  // Don't waste more time looking through cards when we know we have both
  if (hasArchived & hasUnarchived) {
    break;
  }
}

archiveItem.setVisible(hasUnarchived);
unarchiveItem.setVisible(hasArchived);
```
5. The overflow menu on a card doesn't correctly colour for bright backgrounds like other icons do
![image](https://user-images.githubusercontent.com/1885159/173185720-cf1255b2-2f7b-4949-aaf9-1763b9a7dd9a.png)
6. It is a bit weird that you can archive/unarchive both from the overview and from each individual card, but you can only favourite/unfavourite from each individual card. This isn't really an issue with your code, but I am writing it down to not forget that it needs fixing (you're free to add it, otherwise I'll do it later)

If this all is solved I'll go back to nitpicking code, but I think this feature is getting pretty close to ready for release :)

Altonss commented:
> When you archive a card, it disappears completely. This is going to confuse some people. I'd much prefer if "show archived" in "Sort by" defaults to true

Well I don't fully understand this point of view: for me if someone wants to archive a card, he probably thinks that this card will end up in a separate "place" (somewhere a little bit hidden by default). And to just display them by default makes not  much sense to me, because the whole point of archiving is to remove from your view unused cards (that you want to keep archived somewhere).

TheLastProject commented:
> > When you archive a card, it disappears completely. This is going to confuse some people. I'd much prefer if "show archived" in "Sort by" defaults to true
> 
> Well I don't fully understand this point of view: for me if someone wants to archive a card, he probably thinks that this card will end up in a separate "place" (somewhere a little bit hidden by default). And to just display them by default makes not much sense to me, because the whole point of archiving is to remove from your view unused cards (that you want to keep archived somewhere).

I get your point, but if maintaining an app for several years has taught me anything it is that users get confused very easily. Add to that the fact that showing or hiding archived cards is under "Sort by" (which honestly makes no sense, but I am not sure where to better put it), people will have difficulty figuring out how to hide their archived cards. By showing them by default (but at the bottom of the list, see point 3), we can make sure that users know how to show/hide archived cards before they can't find them anymore

TheLastProject commented:
Although maybe it would be way better to have "Archived" be a button like "Groups"

TheLastProject commented:
If Archive becomes a separate thing like Groups, we also don't even need an archive icon and #2 could be skipped.

I understand that a change this big this late in development can be annoying though, so if you're like "ugh I don't want to code that" I can take over

Jasielprogramador commented:
I can do it don't worry but i will need some time

Jasielprogramador commented:
Ok the thing is that when i started to do this issue you told me this:

Now that I'm thinking about this more, creating a new "Archived" tab may actually not be the best solution, because then it becomes hard to see which group an archived card belonged to.

So, i don't really know what is the best thing because you have the group thing, this thing is a little bit tricky to be honest

Altonss commented:
> Ok the thing is that when i started to do this issue you told me this:
> 
> Now that I'm thinking about this more, creating a new "Archived" tab may actually not be the best solution, because then it becomes hard to see which group an archived card belonged to.
> 
> So, i don't really know what is the best thing because you have the group thing, this thing is a little bit tricky to be honest

Yes you're right, archiving also seems tricky now that you are raising this group issue... Hmmm there is also the fact that an archived card may belong to a group that has since been deleted.
I have ideas. The first complicated and the second very simple (and maybe the best/clearest):

1. Duplicate the functionning of how the current cards are handled (with starred/non starred and groups (in tabs)). So we would have this interface started when you click on the "archived' button. This archived interface would look similar to the normal view (non archived) but with a clear distinction that shows "this is the archived interface".
2. Just have an archived button in the menu (like @TheLastProject proposed) where you have access to a raw list of cards (ordered alphabetically for example, or by archiving date) without ANYTHING more ;) (no starred status/group belonging...). This would be straight forward for the user and once you unarchive a card it simply appears back in the main list (the user decides than if he wants to add back a star/group). (and if needed we could improve this with group/star status handling, but with keeping it simple and not exposing too much to the user)

Just my few thoughts, hope it might help to make the best desing decision possible :)

TheLastProject commented:
I'd think option 1 would be easiest. May just actually stay at the main activity but rename "Catima" to "Archive" and have the back button go back to Main.

> Hmmm there is also the fact that an archived card may belong to a group that has since been deleted.

I'm not sure if it would be worth the effort to support this. It would probably be better to just hide empty groups from the main list so you can keep groups that only contain archived cards but don't have them clutter the main screen.

Altonss commented:
> It would probably be better to just hide empty groups from the main list so you can keep groups that only contain archived cards but don't have them clutter the main screen.

I'm not sure about this, because having an empty group is not something "wrong" or impossible in the eyes of users. And to just hide an empty group would make feel the user that is has disappeard... Maybe an "archive" group could be useful :laughing:

Jasielprogramador commented:
Hello there, I'm having lots of problems to reproduce and repair this task:
The overflow menu on a card doesn't correctly colour for bright backgrounds like other icons do

I don't know why but my phone doesn't reproduce the error but the virtual phone in android studio does so I'm pretty lost

TheLastProject commented:
The black/white colours are generated based on the card's colour. Maybe try explicitly picking black as colour for the loyalty card icon: Edit Card -> Tap the Letter left of the name -> Select color -> Black.

If it is still different then maybe your phone and emulator run different Android versions and it's an Android-version-specific issue and you'll have to make sure your emulator runs an affected Android release.

Jasielprogramador commented:
Hello well i have some issues:
1) The colors thing is a little bit of a mess, is there a way like using something like this 
```[java]
menu.findItem(R.id.action_share).setIcon(getIcon(R.drawable.ic_share_white, backgroundNeedsDarkIcons));
```
or some way to change the color of the overflow menu, i saw a way using the  <styles> but it did not work

2)

 I'd think option 1 would be easiest. May just actually stay at the main activity but rename "Catima" to "Archive" and have the back button go back to Main.

Here i understand you mean to create a new "window" where the archived cards are shown, that will have this:

![WhatsApp Image 2022-06-12 at 2 37 49 PM](https://user-images.githubusercontent.com/49266452/173233801-437378e0-b0bc-493d-94bd-233083b4451d.jpeg)

If that is what you meant i have some questions:
1- When you archive a card the window is changed to this "archived window"
2- How can you see the archived cards without archiving one and then changing to the other window
3- If i implement this 2 things i don't know if i should remove the checkbox that if pressed shows all the cards archived ones also
               
If i didn't get something right please tell me because i don't really get this

TheLastProject commented:
> 1. The colors thing is a little bit of a mess, is there a way like using something like this
> 
> 
> ```
> menu.findItem(R.id.action_share).setIcon(getIcon(R.drawable.ic_share_white, backgroundNeedsDarkIcons));
> ```
> 
> or some way to change the color of the overflow menu, i saw a way using the but it did not work

I'm not sure either, never did that before. I found this online but not tested: https://stackoverflow.com/a/51908890

>  1- When you archive a card the window is changed to this "archived window"

Yup, that's a simple `setTitle` call.

>  2- How can you see the archived cards without archiving one and then changing to the other window

You can't, archived cards will only be shown when you go to your archive.

>  3- If i implement this 2 things i don't know if i should remove the checkbox that if pressed shows all the cards archived ones also

Would be better to remove the checkbox yeah, otherwise viewing archived cards would be in 2 places which could be confusing.

Jasielprogramador commented:
You can't, archived cards will only be shown when you go to your archive. 

Yeah, but i mean if the user wants to unarchive a archived card how can he see all the archived cards.

TheLastProject commented:
> Yeah, but i mean if the user wants to unarchive a archived card how can he see all the archived cards.

He opens Catima, clicks the 3 dots in the right top and clicks "Archive". Then all archived cards are shown instead of all non-archived cards.

Jasielprogramador commented:
Sorry for bothering but i have a lot of questions about this issue:
So the thing is that, this new window will ONLY show archived cards with his groups and all that, basically as the xml of the MainActivity class but only with archived cards, then when you press the back button you will be sent to the main page where ONLY the unarchived cards are shown.

Then, how can the user enter this new window without archiving a card ( f.e. he wants to see the archived cards that he has right now), because if you want to do this i image the user should be able to change to this archived windows whenever he wants

The easiest way to do this i think it would be to create a button or whatever to change to this new window (that isn't really a new window is just the MainActivity but some conditions will only show the archived cards)

TheLastProject commented:
> So the thing is that, this new window will ONLY show archived cards with his groups and all that, basically as the xml of the MainActivity class but only with archived cards, then when you press the back button you will be sent to the main page where ONLY the unarchived cards are shown.

Yes.

> Then, how can the user enter this new window without archiving a card ( f.e. he wants to see the archived cards that he has right now), because if you want to do this i image the user should be able to change to this archived windows whenever he wants

They click Archive in the top right, to get to their archive. Not archive on a card:

![image](https://user-images.githubusercontent.com/1885159/173237532-810dfda5-6307-4c3b-a540-c64d0a862686.png)

> The easiest way to do this i think it would be to create a button or whatever to change to this new window (that isn't really a new window is just the MainActivity but some conditions will only show the archived cards)

Yes exactly

Jasielprogramador commented:
Well i think everything is good to go now, i changed some things by myself cause they made no sense, anyway i think this is the easiest solution and finally meets every requirement that was asked for in the issue

TheLastProject commented:
While this works great, I really do not want the MainActivity copy-pasted to a new file because then the same code is in 2 places and there will become a serious risk of something being changed in only 1 file and subtle bugs getting introduced. I think it would be much better to have a variable called something like `mArchiveMode` and if it's true the archive is shown and if it's false the normal cards are shown. Then you can enable `mArchiveMode` when someone presses the `Archive` button and disable it when someone presses the back button (see `onBackPressed` in `MainActivity`).

I'm also seeing some other issues:
1. There is still a "Show archived cards" menu in the sort menu (even though it does nothing anymore)
2. The favourites icon isn't shown if an archived card is favourited (you could probably just remove the archive icon now that archived cards are always in another screen)
3. Selecting multiple cards always shows both "Add to favorites" and "Remove from favorites". Code similar to point 4 in https://github.com/CatimaLoyalty/Android/pull/885#issuecomment-1152902368 should help to only show the relevant button

Still, very cool to see this progress!

Jasielprogramador commented:
Hello there, i have some issues about what you commented:
1) You told me that the favourites icon isn't shown if archived but that's not true as seen here:

https://user-images.githubusercontent.com/49266452/173420906-b9964c24-088e-4016-ba20-a21d3224cc01.mp4

Also i think it would be helpful for the user to leave the archived icon in these cards, this is a visual aid that can be used by the user to remember in which screen is located and what are these cards. I think that  helps more having the icons, because personally it was helpful to have these archived icons to see what screen i was at the moment while testing

2) 
```[text]
 I think it would be much better to have a variable called something like mArchiveMode and if it's true the archive is shown and if it's false the normal cards are shown. Then you can enable mArchiveMode when someone presses the Archive button and disable it when someone presses the back button (see onBackPressed in MainActivity)
```
I think the best way to do this would be to pass a bundle atribute to the MainActivity class, with the value of mArchiveMode, this way a new instance of MainActivity is created, i tried to do it with the onBackPressed method but it does not work

TheLastProject commented:
> 1. You told me that the favourites icon isn't shown if archived but that's not true as seen here:

Oops, you're right, I got confused when switching around screens too much. Sorry!

> Also i think it would be helpful for the user to leave the archived icon in these cards, this is a visual aid that can be used by the user to remember in which screen is located and what are these cards. I think that helps more having the icons, because personally it was helpful to have these archived icons to see what screen i was at the moment while testing

Fair enough!

> I think the best way to do this would be to pass a bundle atribute to the MainActivity class, with the value of mArchiveMode, this way a new instance of MainActivity is created, i tried to do it with the onBackPressed method but it does not work

That sounds even better yeah, I completely forgot that's something Android allows you to do. Nice one!

Jasielprogramador commented:
Well i hope everything is good to go now

TheLastProject commented:
I think we still need a back button in the top left on archive and the archive icon would need to do the 2 color trick the star uses for better contrast on some colors (in this yellow example the white icon is almost impossible to see):

![image](https://user-images.githubusercontent.com/1885159/173437008-248fd40a-de06-420f-9472-b810f4b69c0a.png)

The back button seems to be enabled by using:
```
getSupportActionBar().setDisplayHomeAsUpEnabled(true);
```
and then made working by adding to `onOptionsItemSelected`
```
if (id == android.R.id.home) {
    onBackPressed();
}
```

I'd also like the Archive option in the overflow menu to be nearer to the top (probably as first entry, maybe after Groups) and hidden if you don't have any archived cards (to keep the menu cleaner).

For checking if archived cards exist it is probably easiest to make a new function in `DBHelper` like:
```
public static int getArchivedCardsCount(SQLiteDatabase database) {
    return (int) DatabaseUtils.queryNumEntries(database, LoyaltyCardDbIds.TABLE,
            whereAttrs(LoyaltyCardDbIds.CARD_ID), withArgs(1));
}
```
(untested, based on `getGroupCardCount`)

After that I just need to go through the code again to make sure it's clean, but functionality wise it seems great! The flow is really nice now, useful but out of the way!

Jasielprogramador commented:
Hi @TheLastProject  could you please cancel this workflow:
https://github.com/CatimaLoyalty/Android/actions/runs/2493253330

Jasielprogramador commented:
Ok i think everything is good to go now

TheLastProject commented on app/src/test/java/protect/card_locker/ImportURITest.java:
```diff
@@ -70,7 +70,7 @@ public void ensureNoDataLoss() throws InvalidObjectException, UnsupportedEncodin
     public void ensureNoCrashOnMissingHeaderFields() throws InvalidObjectException, UnsupportedEncodingException {
         // Generate card
-        DBHelper.insertLoyaltyCard(mDatabase, "store", "note", null, new BigDecimal("10.00"), Currency.getInstance("EUR"), BarcodeFormat.UPC_A.toString(), null, CatimaBarcode.fromBarcode(BarcodeFormat.QR_CODE), null, 0, null);
+        DBHelper.insertLoyaltyCard(mDatabase, "store", "note", null, new BigDecimal("10.00"), Currency.getInstance("EUR"), BarcodeFormat.UPC_A.toString(), null, CatimaBarcode.fromBarcode(BarcodeFormat.QR_CODE), null, 0, null,0);
```
Please also test if the value is correct in // Compare everything

In this case it should be 0 because if someone shares a card with us, there is no data if that user had it archived.

TheLastProject commented on app/src/test/java/protect/card_locker/LoyaltyCardCursorAdapterTest.java:
```diff
@@ -162,9 +162,9 @@ public void TestCursorAdapterFontSizes() {
 
     @Test
     public void TestCursorAdapterStarring() {
```
This test already checks `assertEquals(View.VISIBLE, star.getVisibility());`, maybe we can test `assertEquals(View.VISIBLE, archiveIcon.getVisibility());` here too so we have archive tests too (or whatever the archive icon is called, I forgot)

TheLastProject commented on app/src/main/java/protect/card_locker/MainActivity.java:
```diff
@@ -175,17 +221,33 @@ public void onDestroyActionMode(ActionMode inputMode) {
+            setContentView(R.layout.main_activity);
+        }
+        else{
+            setTitle(R.string.archive);
```
In English the action and the list are the same, but that's not necessarily the case for other languages. So we need to split this up in 2 strings.

|  | English | Dutch |
| - | - | - |
| Move to archive | Archive | Archiveer |
| Archived cards | Archive | Archief |

```suggestion
            setTitle(R.string.archiveList);
```

TheLastProject commented on app/build.gradle:
```diff
@@ -76,6 +76,9 @@ android {
     }
+    lintOptions {
+        abortOnError false
+    }
```
Please remove your change here :)

TheLastProject reviewed (changes requested):
Looking great! Just small last fixes!

Jasielprogramador commented:
Oh yeah the bug was just a silly thing , basically all these actions had the tag app:showAsAction="ifRoom"/> and now they have app:showAsAction="never"/>

Jasielprogramador commented:
Well i also added this method in DBHelper class
```[java]
    public static Cursor getAllLoyaltyCardCursor(SQLiteDatabase database){
        return database.query(LoyaltyCardDbIds.TABLE,null,null,null,null,null,null);
    }
```

i implemented that because of this method found in LoyaltyCardCursorAdapterTest class:

```[java]

 @Test
    public void TestCursorAdapterStarring() {
        assertNotEquals(-1, DBHelper.insertLoyaltyCard(mDatabase, "storeA", "note", null, new BigDecimal("0"), null, "cardId", null, CatimaBarcode.fromBarcode(BarcodeFormat.UPC_A), Color.BLACK, 0, null,0));
        assertNotEquals(-1, DBHelper.insertLoyaltyCard(mDatabase, "storeB", "note", null, new BigDecimal("0"), null, "cardId", null, CatimaBarcode.fromBarcode(BarcodeFormat.UPC_A), Color.BLACK, 1, null,1));
        assertNotEquals(-1, DBHelper.insertLoyaltyCard(mDatabase, "storeC", "note", null, new BigDecimal("0"), null, "cardId", null, CatimaBarcode.fromBarcode(BarcodeFormat.UPC_A), Color.BLACK, 1, null,1));

        assertEquals(3, DBHelper.getLoyaltyCardCount(mDatabase));




        Cursor cursor = DBHelper.getAllLoyaltyCardCursor(mDatabase);

```

Right now the getLoyaltyCardCursor only gives you by default all the unarchived cards, this is done this way because in the main screen you will only see these ones . So i created this other method, for me it was the easier way to do it i dunno if there's a easier way, and also btw i tested the visibility of the archived icon in this same method

Jasielprogramador commented:
Sorry i did not understand what you meant with the bug yeah i'm changing that rn dw

TheLastProject commented:
Merged through #902, thank you a lot!
